tripal/tripal
# CV terms on all fields

# Tripal 4 Core Dev Task

### Issue #1789


### Tripal Version: 4.x

## Description


Currently, users can only set CV terms on Tripal fields. But someone can attach an existing Drupal field to a Tripal content type. For example, if a user wants to add an image to an organism content type they can do so using the Drupal image field, but, that image field doesn't have a CV term associated with it. For web services to work correctly all fields attached to a Tripal content type must be associated with a CV term.

This PR fixes the problem by using a `hook_form_alter()` function to add the CV term form elements to any field settings page that is attached to a tripal entity type.    It also sets a 3rd party setting with the field so that later, when we setup web services, we can get the CV term from the field.  

## Testing?



1. On an existing Tripal site, go to admin > Structure > Tripal Content Types and select "manage fields" for the "Organism" content type.
2. Click the 'Create a New Field' button and in the page that appears, add an Image field (Drupal 10.2 first select "File upload"). Name it "Organism Image".
3. When editing the image you will first see the field settings page.  Click the "save field settings button"
4. You will recieve the warning:  "The field is missing an assigned controlled vocabulary term. Please set one"
5. Scroll to the section "Controlled Vocabulary Section" and add the term "Image (data:2968)"
6. Click the "save settings" button at the bottom.  You now have a new image field. 
7. Click the link to edit the new "Organism Image" field. You will see that in the "Controlled Vocabulary" fieldset that the term is set and present.  You can change the term, save, and re-open and the term change is maintained.
